### PR TITLE
Creating the context object [WPB-10915]

### PR DESCRIPTION
**Pull request by augustocdias**
# What's new in this PR

This PR introduces the Context object, which is responsible for providing transactional support for Core Crypto operations. Clients will be able to batch operations and persist in a single call. All operations that persist data were moved to this new Context object. A new Mls Crypto Provider was introduced but using the transaction as the key store instead of the DB connection. All other changes consist in moving operations into the context from the MlsCentral and using the new mls provider instead.

----
##### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: crypto/src/error.rs [hunk 1]
```diff
@@ -249,6 +249,9 @@ pub enum CryptoError {
     /// Not supported for the moment
     #[error("Not supported for the moment")]
     Unsupported,
+    /// Invalid Context. This context was finished and cannot be further used
+    #[error("This context was already finished and cannot be further used")]
```
Comment (SimonThormeyer) [inline crypto/src/error.rs:253]: "cannot be further used" -> "can no longer be used" (More natural phrasing)
```suggestion
    /// Invalid Context. This context has been finished and can no longer be used.
    #[error("This context has already been finished and can no longer be used.")]
```

File: crypto/src/mls/client/id.rs [hunk 2]
```diff
@@ -16,6 +16,8 @@
 
 use crate::CryptoError;
 
+/// A Client identifier
+///
```
Comment (SimonThormeyer) [inline crypto/src/mls/client/id.rs:20]: Is this intentional? It seems incomplete

File: crypto/src/mls/context.rs [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+    }
+
+    pub(crate) async fn mls_client(&self) -> CryptoResult<RwLockReadGuardArc<Option<Client>>> {
+        match self.state.read().await.deref() {
+            ContextState::Valid {
+                mls_client,
+                callbacks: _,
+                transaction: _,
+                mls_groups: _,
+            } => Ok(mls_client.read_arc().await),
```
Comment (SimonThormeyer) [inline crypto/src/mls/context.rs:81]: We can use `..` for unused fields
```suggestion
            ContextState::Valid { mls_client, .. } => Ok(mls_client.read_arc().await),
```

File: crypto/src/mls/context.rs [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+    }
+
+    pub(crate) async fn mls_client_mut(&self) -> CryptoResult<RwLockWriteGuardArc<Option<Client>>> {
+        match self.state.read().await.deref() {
+            ContextState::Valid {
+                mls_client,
+                callbacks: _,
+                transaction: _,
+                mls_groups: _,
+            } => Ok(mls_client.write_arc().await),
```
Comment (SimonThormeyer) [inline crypto/src/mls/context.rs:93]: See above

File: crypto/src/mls/context.rs [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+
+    pub(crate) async fn callbacks(
+        &self,
+    ) -> CryptoResult<RwLockReadGuardArc<Option<std::sync::Arc<dyn CoreCryptoCallbacks + 'static>>>> {
+        match self.state.read().await.deref() {
+            ContextState::Valid {
+                mls_client: _,
+                callbacks,
+                transaction: _,
+                mls_groups: _,
```
Comment (SimonThormeyer) [inline crypto/src/mls/context.rs:106]: See above

File: crypto/src/mls/context.rs [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+    }
+
+    pub(crate) async fn mls_provider(&self) -> CryptoResult<TransactionalCryptoProvider> {
+        match self.state.read().await.deref() {
+            ContextState::Valid {
+                mls_client: _,
+                callbacks: _,
+                transaction,
+                mls_groups: _,
+            } => Ok(transaction.clone()),
```
Comment (SimonThormeyer) [inline crypto/src/mls/context.rs:119]: See above

File: crypto/src/mls/context.rs [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+    }
+
+    pub(crate) async fn transaction(&self) -> CryptoResult<KeystoreTransaction> {
+        match self.state.read().await.deref() {
+            ContextState::Valid {
+                mls_client: _,
+                callbacks: _,
+                transaction,
+                mls_groups: _,
+            } => Ok(transaction.transaction()),
```
Comment (SimonThormeyer) [inline crypto/src/mls/context.rs:131]: See above

File: keystore/src/entities/platform/wasm/mls/credential.rs [hunk 8]
```diff
@@ -118,13 +109,13 @@ impl MlsCredentialExt for MlsCredential {
...
                 let mut credential = serde_wasm_bindgen::from_value::<MlsCredential>(entity_raw)?;
-                credential.decrypt(&storage.cipher)?;
+                credential.decrypt(&cipher)?;
 
                 let id = JsValue::from(credential.id.clone());
                 let request = store.delete(id)?;
                 request.await?;
             }
-            WasmStorageWrapper::InMemory(_) => {
+            WasmStorageTransaction::InMemory { db: _, cipher: _ } => {
```
Comment (SimonThormeyer) [inline keystore/src/entities/platform/wasm/mls/credential.rs:118]:
```suggestion
            WasmStorageTransaction::InMemory { .. } => {
```

**Review by SimonThormeyer — APPROVED** [review]
Looks good! Just some minor things

File: crypto/src/mls/client/id.rs (hunk 2 above)
Comment (augustocdias) [reply crypto/src/mls/client/id.rs:20]: clippy was complaining about paragraphs being too long... I just added something resembling a title

**Pull request merged**